### ISSUE TITLE
Prometheus replying with 403 after update to 2.13

### ISSUE THREAD

**Issue by eugeneromero** [linkerd/linkerd2#10790]
### What is the issue?

After updating the `linkerd-viz` chart to versions `30.8.0` or `30.8.1` (which update the components to version `2.13.0` and `2.13.1` respectively), we noticed that all our canaries started failing. Looking at the Flagger pod logs, we see this as the cause of the failing checks:
```
{"level":"error","ts":"2023-04-19T18:03:36.443Z","caller":"controller/events.go:39","msg":"Prometheus query failed: running query failed: error response: ","canary":"my-canary","stacktrace":"github.com/fluxcd/flagger/pkg/controller.(*Controller).recordEventErrorf\n\t/workspace/pkg/controller/events.go:39\ngithub.com/fluxcd/flagger/pkg/controller.(*Controller).runBuiltinMetricChecks\n\t/workspace/pkg/controller/scheduler_metrics.go:145\ngithub.com/fluxcd/flagger/pkg/controller.(*Controller).runAnalysis\n\t/workspace/pkg/controller/scheduler.go:716\ngithub.com/fluxcd/flagger/pkg/controller.(*Controller).advanceCanary\n\t/workspace/pkg/controller/scheduler.go:413\ngithub.com/fluxcd/flagger/pkg/controller.CanaryJob.Start.func1\n\t/workspace/pkg/controller/job.go:39"}
```

So, it seems that Prometheus is not answering correctly. To verify, we tried calling the Prometheus URL from Flagger (or any other container in the cluster):
```
/ # curl -v http://prometheus.linkerd-viz:9090
*   Trying 10.0.233.97:9090...
* Connected to prometheus.linkerd-viz (10.0.233.97) port 9090 (#0)
> GET / HTTP/1.1
> Host: prometheus.linkerd-viz:9090
> User-Agent: curl/8.0.1
> Accept: */*
>
< HTTP/1.1 403 Forbidden
< content-length: 0
< date: Thu, 20 Apr 2023 08:56:28 GMT
<
* Connection #0 to host prometheus.linkerd-viz left intact```
```

Rolling back to version `30.3.6` of the chart, which installs the `2.12.4` app version, makes Prometheus work correctly:
```
/ # curl -v http://prometheus.linkerd-viz:9090
*   Trying 10.0.233.97:9090...
* Connected to prometheus.linkerd-viz (10.0.233.97) port 9090 (#0)
> GET / HTTP/1.1
> Host: prometheus.linkerd-viz:9090
> User-Agent: curl/8.0.1
> Accept: */*
>
< HTTP/1.1 302 Found
< content-type: text/html; charset=utf-8
< location: /graph
< date: Thu, 20 Apr 2023 08:58:53 GMT
< content-length: 29
<
<a href="/graph">Found</a>.

* Connection #0 to host prometheus.linkerd-viz left intact
```

It appears that something has changed with version 2.13, which blocks connections to Prometheus. Interestingly, Prometheus still works internally (logging into it and curl-ing localhost:9090 works correctly). So Prometheus itself does not seem to be the cause of the issue. Possibly something with the new policies?

### How can it be reproduced?

Either a clean install of, or an upgrade to, the linkerd-viz helm chart version 30.8. No real modifications other than maybe installing in a non-default namespace.

Once installed, easiest check is to call the prometheus service from any other container in the cluster:

```
curl http://prometheus.NAMESPACE:9090
```

### Logs, error output, etc

Any calls to Prometheus fail with:

```
HTTP/1.1 403 Forbidden
```

### output of `linkerd check -o short`

```
$ linkerd check -o short
Status check results are √
```

### Environment

- Kubernetes version: 1.25.5
- Environment: AKS
- OS: Linux
- Linkerd version: 2.13

### Would you like to work on fixing this bug?

no

**Comment by alpeb** [comment]
Thanks for the report @eugeneromero. Linkerd has now locked down the access to Prometheus to just the `metrics-api` pod. You can verify that like so:
```
$ linkerd authz -n linkerd-viz po/prometheus-cf44f464b-tml5b
ROUTE   SERVER            AUTHORIZATION_POLICY   SERVER_AUTHORIZATION
*       prometheus-admin  prometheus-admin

$ k -n linkerd-viz get authorizationpolicies.policy.linkerd.io prometheus-admin -oyaml
apiVersion: policy.linkerd.io/v1alpha1
kind: AuthorizationPolicy
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"policy.linkerd.io/v1alpha1","kind":"AuthorizationPolicy","metadata":{"annotations":{"linkerd.io/created-by":"linkerd/helm stable-2.13.1"},"labels":{"linkerd.io/extension":"viz"},"name":"prometheus-admin","namespace":"linkerd-viz"},"spec":{"requiredAuthenticationRefs":[{"kind":"ServiceAccount","name":"metrics-api","namespace":"linkerd-viz"}],"targetRef":{"group":"policy.linkerd.io","kind":"Server","name":"prometheus-admin"}}}
    linkerd.io/created-by: linkerd/helm stable-2.13.1
  creationTimestamp: "2023-04-20T14:18:48Z"
  generation: 1
  labels:
    linkerd.io/extension: viz
  name: prometheus-admin
  namespace: linkerd-viz
  resourceVersion: "1264"
  uid: 09f8b704-7398-4d04-b1e8-50cd66f8e569
spec:
  requiredAuthenticationRefs:
  - kind: ServiceAccount
    name: metrics-api
    namespace: linkerd-viz
  targetRef:
    group: policy.linkerd.io
    kind: Server
    name: prometheus-admin
```

I'll update the Flagger docs with the AuthorizationPolicy that needs to be added, and point to the update here.

**Comment by eugene-romero-capgemini** [comment]
Thank you @alpeb, I appreciate it!

**Comment by alpeb** [comment]
I ended up pushing a PR to the Flagger repo directly: fluxcd/flagger#1417
If you're following our [docs](https://linkerd.io/2.13/tasks/flagger/index.html) and installing via kustomize, you can in the meantime apply from my fork:
```
kubectl apply -k github.com/alpeb/flagger/kustomize/linkerd?ref=linkerd-2.13
```
If you're using a different method, just make sure to apply this AuthorizationPolicy:
```yaml
apiVersion: policy.linkerd.io/v1alpha1
kind: AuthorizationPolicy
metadata:
  namespace: linkerd-viz
  name: prometheus-admin-flagger
spec:
  targetRef:
    group: policy.linkerd.io
    kind: Server
    name: prometheus-admin
  requiredAuthenticationRefs:
    - kind: ServiceAccount
      name: flagger
      namespace: linkerd-flagger
```
assuming linkerd Viz is installed under the `linkerd-viz` namespace, and flagger under the `linkerd-flagger` namespace. Note that flagger's workload needs to be injected by linkerd in order to provide a proper identity and use this auth.

**Comment by eugeneromero** [comment]
Hey @alpeb,

This does indeed fix the issue. Thank you for your work on it!

Since we do not use Kustomize, the way I applied it was:
1. Create new namespace `linkerd-flagger`, ensuring it has the `linkerd.io/inject: enabled` annotation
1. Move Flagger Helm chart to this new namespace
1. Add the new Authorization Policy from your comment above

After this, canaries work again as expected. Thank you again! Feel free to close as soon as your PR is merged.

**Comment by alpeb** [comment]
The change has just been merged in fluxcd/flagger#1417
Note that flagger will get installed into the (inejcted) `flagger-system` namespace :wink: (see PR for more details).
